### PR TITLE
Processing exps

### PR DESCRIPTION
**Pull request by volaya**
Support for expressions in Processing parameter dialogs

### CODE REVIEW

File: python/plugins/processing/gui/AlgorithmDialog.py [hunk 1]
```diff
@@ -121,6 +121,18 @@ def setParamValues(self):
 
         return True
 
+    def evaluateExpression(self, text):
+        context = QgsExpressionContext()
+        context.appendScope(QgsExpressionContextUtils.projectScope())
```
Comment (nyalldawson) [inline python/plugins/processing/gui/AlgorithmDialog.py]: you should first add the global scope:
context.appendScope(QgsExpressionContextUtils.globalScope())

File: python/plugins/processing/gui/NumberInputPanel.py [hunk 2]
```diff
@@ -74,15 +80,64 @@ def __init__(self, number, minimum, maximum, isInteger):
...
-    def showNumberInputDialog(self):
-        dlg = NumberInputDialog(self.isInteger)
-        dlg.exec_()
-        if dlg.value is not None:
-            self.spnValue.setValue(dlg.value)
+    def showExpressionsBuilder(self):
+        context = QgsExpressionContext()
+        context.appendScope(QgsExpressionContextUtils.globalScope())
+        scope = QgsExpressionContextUtils.projectScope()
+        self.addVariablesToScope(scope)
```
Comment (nyalldawson) [inline python/plugins/processing/gui/NumberInputPanel.py]: Strictly speaking you should add the projectScope intact, then add a new blank scope for the processing variables.

File: python/plugins/processing/gui/NumberInputPanel.py [hunk 3]
```diff
@@ -74,15 +80,64 @@ def __init__(self, number, minimum, maximum, isInteger):
...
+        for layer in layers:
+            for i in range(layer.bandCount()):
+                stats = layer.dataProvider().bandStatistics(i + 1)
+                scope.setVariable("%s_band%i_avg" % (name, i + 1), stats.mean)
+                scope.setVariable("%s_band%i_stddev" % (name, i + 1), stats.stdDev)
+                scope.setVariable("%s_band%i_min" % (name, i + 1), stats.minimumValue)
+                scope.setVariable("%s_band%i_max" % (name, i + 1), stats.maximumValue)
+
+        extent = iface.mapCanvas().extent()
+        scope.setVariable("canvasextent_minx", extent.xMinimum())
```
Comment (nyalldawson) [inline python/plugins/processing/gui/NumberInputPanel.py]: All these new variables should have a description added in QgsExpression::initVariableHelp()

File: python/plugins/processing/gui/OutputSelectionPanel.py [hunk 4]
```diff
@@ -100,6 +106,20 @@ def selectOutput(self):
 
             popupMenu.exec_(QCursor.pos())
 
+    def showExpressionsBuilder(self):
+        context = QgsExpressionContext()
+        context.appendScope(QgsExpressionContextUtils.globalScope())
+        scope = QgsExpressionContextUtils.projectScope()
+        self.addVariablesToScope(scope)
```
Comment (nyalldawson) [inline python/plugins/processing/gui/OutputSelectionPanel.py]: same here - this should add the projectScope intact, then create a new blank scope for the processing variables

**Comment by nyalldawson** [comment]
Looks good, with a few fixes to how variables and scopes should be handled.

On a related note, can something be done about https://github.com/qgis/QGIS/commit/794ef725202825fa71d80d4733d5f554c4f8c4a9 ? As the comments say, that commit has broken lots of things....

**Comment by volaya** [comment]
@nyalldawson Thanks for the comments! I tried to figure out how expressiones worked, but help is really welcome. I will fix it and update the PR

Regarding the other issue, I wasn't aware of it (been on vacation the last 2 weeks...) but you are right, the replacing is a bit too "wild". I will try to fine tune that to avoid breaking scripts. 

Thanks!

File: python/plugins/processing/gui/NumberInputPanel.py (hunk 3 above)
Comment (volaya) [inline python/plugins/processing/gui/NumberInputPanel.py]: @nyalldawson that method is not available from python.

https://qgis.org/api/classQgsExpression.html#abd76aaa8d7416facf985bcc0306e8404

 Any other way of adding the variable description?

**Comment by alexbruy** [comment]
Manually merged

**Pull request merged**